pkg/storage/s3.go

```diff
@@ -1,56 +1,318 @@
 		return
 	}
-
 	options.Region, err = flags.GetString(s3RegionOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
+	options.SSE, err = flags.GetString(s3SSEOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ACL, err = flags.GetString(s3ACLOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.StorageClass, err = flags.GetString(s3StorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ForcePathStyle = true
+	options.Provider, err = flags.GetString(s3ProviderOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	return options, err
+}
+
+// newS3Storage initialize a new s3 storage for metadata
+func newS3Storage(s3Back *backup.S3) (*S3Storage, error) {
+	qs := *s3Back
+	v := credentials.Value{
+		AccessKeyID:     qs.AccessKey,
+		SecretAccessKey: qs.SecretAccessKey,
+	}
+
+	// low timeout to ec2 metadata service
+	lowTimeoutClient := &http.Client{Timeout: 1 * time.Second}
+	def := defaults.Get()
+	def.Config.HTTPClient = lowTimeoutClient
+	ec2Session, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	// first provider to supply a credential set "wins"
+	providers := []credentials.Provider{
+		// use static credentials if they're present (checked by provider)
+		&credentials.StaticProvider{Value: v},
+
+		// * Access Key ID:     AWS_ACCESS_KEY_ID or AWS_ACCESS_KEY
+		// * Secret Access Key: AWS_SECRET_ACCESS_KEY or AWS_SECRET_KEY
+		&credentials.EnvProvider{},
+
+		// A SharedCredentialsProvider retrieves credentials
+		// from the current user's home directory.  It checks
+		// AWS_SHARED_CREDENTIALS_FILE and AWS_PROFILE too.
+		&credentials.SharedCredentialsProvider{},
+
+		// Pick up IAM role if we're in an ECS task
+		defaults.RemoteCredProvider(*def.Config, def.Handlers),
+
+		// Pick up IAM role in case we're on EC2
+		&ec2rolecreds.EC2RoleProvider{
+			Client: ec2metadata.New(ec2Session, &aws.Config{
+				HTTPClient: lowTimeoutClient,
+			}),
+			ExpiryWindow: 3 * time.Minute,
+		},
+	}
+	cred := credentials.NewChainCredentials(providers)
 
-	// TODO: Add more options here.
+	if qs.Region == "" && qs.Endpoint == "" {
+		qs.Endpoint = "https://s3.amazonaws.com/"
+	}
+	if qs.Region == "" {
+		qs.Region = "us-east-1"
+	}
 
-	return
+	awsConfig := aws.NewConfig().
+		WithMaxRetries(maxRetries).
+		WithCredentials(cred).
+		WithS3ForcePathStyle(qs.ForcePathStyle)
+	if qs.Region != "" {
+		awsConfig.WithRegion(qs.Region)
+	}
+	if qs.Endpoint != "" {
+		awsConfig.WithEndpoint(qs.Endpoint)
+	}
+	// awsConfig.WithLogLevel(aws.LogDebugWithSigning)
+	awsSessionOpts := session.Options{
+		Config: *awsConfig,
+	}
+	ses, err := session.NewSessionWithOptions(awsSessionOpts)
+	if err != nil {
+		return nil, err
+	}
+	c := s3.New(ses)
+	err = checkS3Bucket(c, qs.Bucket)
+	if err != nil {
+		return nil, errors.Errorf("checkS3Bucket error: %v", err)
+	}
+
+	qs.Prefix = strings.Trim(qs.Prefix, "/")
+	qs.Prefix += "/"
```

Comment:
Let's move these into `apply()`.

Reply:
The prefix is not in options struct.

Reply:
it is in `*backup.S3`

Reply:
`apply` is to apply the options to backup.S3, it's not reasonable to change the fields that are not in the options.
BTW, I have trimed the "/" before calling `apply`.

Reply:
Will remove the `Trim` here.

Reply:
done